oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/SampleFamily.java

```diff
@@ -376,15 +374,63 @@ public SampleFamily endpoint(List<String> serviceKeys, List<String> endpointKeys
+        return createMeterSamples(serviceKeys, endpointKeys, ScopeType.ENDPOINT);
+    }
+
+    private SampleFamily createMeterSamples(List<String> serviceKeys, List<String> level2keys, ScopeType scopeType) {
+        final List<String> labelKeys = io.vavr.collection.Stream.concat(serviceKeys, level2keys).asJava();
+        Map<MeterEntity, Sample[]> meterSamples = new HashMap<>();
+        Arrays.stream(samples)
+              .collect(groupingBy(it -> getLabels(labelKeys, it), mapping(identity(), toList())))
+              .forEach((labels, samples) -> {
+                  MeterEntity meterEntity = getMeterEntity(samples, serviceKeys, level2keys, scopeType);
+                  Preconditions.checkNotNull(meterEntity);
+                  meterSamples.put(meterEntity, left(labelKeys, samples));
+              });
+
+        this.context.setMeterSamples(meterSamples);
+        //This samples is original, The grouped samples is in context which mapping with MeterEntity
+        return SampleFamily.build(this.context, samples);
+    }
+
+    private MeterEntity getMeterEntity(List<Sample> samples,
+                                       List<String> serviceKeys,
+                                       List<String> level2keys,
+                                       ScopeType scopeType) {
+        switch (scopeType) {
+            case SERVICE:
+                return MeterEntity.newService(dim(serviceKeys, samples));
+            case SERVICE_INSTANCE:
+                return MeterEntity.newServiceInstance(dim(serviceKeys, samples), dim(level2keys, samples));
+            case ENDPOINT:
+                return MeterEntity.newEndpoint(dim(serviceKeys, samples), dim(level2keys, samples));
+        }
+        return null;
     }
 
     private String dim(List<String> labelKeys) {
-        String name = labelKeys.stream().map(k -> samples[0].labels.getOrDefault(k, "")).collect(Collectors.joining("."));
+        String name = labelKeys.stream()
+                               .map(k -> samples[0].labels.getOrDefault(k, ""))
+                               .collect(Collectors.joining("."));
+        return CharMatcher.is('.').trimFrom(name);
+    }
+
+    private String dim(List<String> labelKeys, List<Sample> samples) {
+        String name = labelKeys.stream()
+                               .map(k -> samples.get(0).labels.getOrDefault(k, ""))
+                               .collect(Collectors.joining("."));
         return CharMatcher.is('.').trimFrom(name);
     }
 
+    private Sample[] left(List<String> labelKeys, List<Sample> samples) {
+        return samples.stream().map(s -> {
+            ImmutableMap<String, String> ll = ImmutableMap.<String, String>builder()
+                .putAll(Maps.filterKeys(s.labels, key -> !labelKeys.contains(key)))
+                .build();
+            return s.toBuilder().labels(ll).build();
+        }).toArray(Sample[]::new);
+    }
+
     private SampleFamily left(List<String> labelKeys) {
```

Comment:
This is never used, if not useful, please remove this.

Reply:
done